### ISSUE TITLE
Android GL failing to create swap chain.

### ISSUE THREAD

**Issue by nmpribeiro** [gfx-rs/wgpu#1757]
**Description**
Android bug on swap chain descriptors. Basically, I got blocked by the following 
`adapter.get_swap_chain_preferred_format(&surface).unwrap();` 
and

**Repro steps**
Triangle example compiled for android (using NativeActivity build).

**Expected vs observed behavior**
I get the following:
```
I/RustStdoutStderr: thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', /Users/<USER>/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-core-0.8.1/src/hub.rs:880:29
```
while I expect passing to the next execution line with a swap chain descriptor created.

**Extra materials**
If I define the `TextureFormat` statically (`Rgba16Float), I can create a SwapChainDescriptor struct but it fails on the next line:
```
log::info!("Swap chain");
let swap_chain = device.create_swap_chain(&surface, &sc_desc); // <- fails here
log::info!("Swap chain done!"); // <- never reached
```
Output:
```
wgpu_core::device: creating swap chain SwapChainDescriptor { usage: RENDER_ATTACHMENT, format: Rgba16Float, width: 1080, height: 2280, present_mode: Fifo }
I/RustStdoutStderr: thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', /Users/<USER>/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-core-0.8.1/src/hub.rs:880:29
```

Unfortunately I was not able to trigger `RUST_BACKTRACE=full`.

**Platform**
rustc 1.56.0-nightly (9c25eb7aa 2021-07-25)
building android for min sdk 16 and target 29
gpu: "Android Emulator OpenGL ES Translator (AMD Radeon Pro 5500M OpenGL Engine)"
ogl: Version: 3.0, (4.1 ATI-4.5.14)
shading lang: Shading Language: 3.0
wgpu: 0.8.1

**Comment by nmpribeiro** [comment]
Also happens with wpgu: 0.9

```
I/hello-world: rust_jsx_app: Swap chain Descriptor
I/RustStdoutStderr: thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', /Users/<USERS>/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-core-0.9.2/src/hub.rs:878:29
```

**Comment by nmpribeiro** [comment]
In an  android I have it works fine with vulkan, but in my simulator it doesn't.

**Comment by kvark** [comment]
Could you try the master branch, please?

**Comment by nmpribeiro** [comment]
@kvark , I have tried to clone this repo, then point `wgpu = { path = "../wgpu_repo" }` but it is telling me it's a virtual project. When I try to point to path `../wgpu_repo/wgpu` a bunch of errors come up. Is there a special way one need to point to this repo to make it work with our projects?

**Comment by kvark** [comment]
Nothing special, except maybe `resolver = "2"`. Errors are expected, because there are breaking changes between v0.8.1 and master.

**Comment by Gordon-F** [comment]
@nmpribeiro Based on stacktrace, surface in your application already dropped. Make sure you handle Suspended/Resume events right.

https://github.com/EmbarkStudios/rust-gpu/blob/cccb9737d723a1be7f1a963a8f57c5874fecf6ff/examples/runners/wgpu/src/graphics.rs#L127-L135

**Comment by nmpribeiro** [comment]
@Gordon-F thanks I will bullet proof those event handlings but fyi, this was on startup of the app. No blurring or focusing of the window happening at all.

**Comment by nmpribeiro** [comment]
@kvark , I'm adding wgpu repo as follows:
```
wgpu = {git = "https://github.com/gfx-rs/wgpu.git"}
```

I get the following, still during startup and State construction (surface creation).

Android simulator:
```
I/System.out: debugger has settled (1324)
W/System: ClassLoader referenced unknown path: 
D/NetworkSecurityConfig: No Network Security Config specified, using platform default
D/NetworkSecurityConfig: No Network Security Config specified, using platform default
D/libEGL: loaded /vendor/lib/egl/libEGL_emulation.so
D/libEGL: loaded /vendor/lib/egl/libGLESv1_CM_emulation.so
D/libEGL: loaded /vendor/lib/egl/libGLESv2_emulation.so
I/hello-world: rust_jsx_app: Will now wait for native window
    rust_jsx_app: Waiting for NativeScreen
I/hello-world: rust_jsx_app: NativeScreen Found:Some(NativeWindow { ptr: 0xe29064d8 })
    rust_jsx_app: Proceeding after native window
    rust_jsx_app: Inside RUN!
    rust_jsx_app: Inner Size: 0 x 0
    rust_jsx_app: Outer Size: 0 x 0
D/HostConnection: HostConnection::get() New Host Connection established 0xe7f62ad0, tid 31826
D/HostConnection: HostComposition ext ANDROID_EMU_CHECKSUM_HELPER_v1 ANDROID_EMU_native_sync_v2 ANDROID_EMU_native_sync_v3 ANDROID_EMU_native_sync_v4 ANDROID_EMU_dma_v1 ANDROID_EMU_direct_mem ANDROID_EMU_host_composition_v1 ANDROID_EMU_host_composition_v2 ANDROID_EMU_YUV_Cache ANDROID_EMU_async_unmap_buffer ANDROID_EMU_has_shared_slots_host_memory_allocator ANDROID_EMU_sync_buffer_data GL_OES_EGL_image_external_essl3 GL_OES_vertex_array_object GL_KHR_texture_compression_astc_ldr ANDROID_EMU_host_side_tracing ANDROID_EMU_async_frame_commands ANDROID_EMU_gles_max_version_3_0 
D/vulkan: searching for layers in '/data/app/~~6yh1lPespYGQcBUTE4u-5g==/io.github.nmpribeiro.androidapp-710BiACbYnep4lorru9VQg==/lib/x86'
    searching for layers in '/data/app/~~6yh1lPespYGQcBUTE4u-5g==/io.github.nmpribeiro.androidapp-710BiACbYnep4lorru9VQg==/base.apk!/lib/x86'
I/hello-world: wgpu_hal::vulkan::instance: Unable to find extension: VK_EXT_debug_utils
    wgpu_hal::vulkan::instance: Unable to find extension: VK_KHR_get_physical_device_properties2
W/hello-world: wgpu_hal::vulkan::instance: Unable to find layer: VK_LAYER_KHRONOS_validation
D/hello-world: wgpu_hal::gles::egl: Client extensions: [
        "EGL_EXT_client_extensions",
        "EGL_KHR_platform_android",
        "EGL_ANGLE_platform_angle",
        "EGL_ANDROID_GLES_layers",
    ]
I/hello-world: wgpu_hal::gles::egl: Using default platform
I/hello-world: wgpu_hal::gles::egl: Display vendor "Android", version (1, 4)
D/hello-world: wgpu_hal::gles::egl: Display extensions: [
        "EGL_KHR_get_all_proc_addresses",
        "EGL_ANDROID_presentation_time",
        "EGL_KHR_swap_buffers_with_damage",
        "EGL_ANDROID_get_native_client_buffer",
        "EGL_ANDROID_front_buffer_auto_refresh",
        "EGL_ANDROID_get_frame_timestamps",
        "EGL_EXT_surface_SMPTE2086_metadata",
        "EGL_EXT_surface_CTA861_3_metadata",
        "EGL_KHR_image",
        "EGL_KHR_image_base",
        "EGL_KHR_gl_texture_2D_image",
        "EGL_KHR_fence_sync",
        "EGL_KHR_create_context",
        "EGL_ANDROID_image_native_buffer",
        "EGL_KHR_wait_sync",
        "EGL_ANDROID_recordable",
    ]
I/hello-world: wgpu_hal::gles::egl: 	Trying presentation
    wgpu_hal::gles::egl: 	EGL context: +debug KHR
W/hello-world: wgpu_hal::gles::egl: 	EGL context: -robust access
D/EGL_emulation: eglCreateContext: 0xe7f63320: maj 3 min 0 rcv 3
W/hello-world: wgpu_hal::gles::egl: 	EGL surface: -srgb
I/hello-world: rust_jsx_app: Surface
I/RustStdoutStderr: thread '<unnamed>' panicked at 'Cannot get the native window, it's null and will always be null before Event::Resumed and after Event::Suspended. Make sure you only call this function between those events.', /Users/nmpribeiro/.cargo/git/checkouts/winit-1e8677f4a4789b6b/70b2bcd/src/platform_impl/android/mod.rs:594:13
    stack backtrace:
I/RustStdoutStderr: note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
I/hello-world: wgpu_core::hub: Dropping Global
W/libEGL: eglTerminate() called w/ 1 objects remaining
```

Physical device:

```
I/System.out: debugger has settled (1336)
W/System: ClassLoader referenced unknown path: 
I/eiro.androidap: QarthPatchMonintor::Init
    QarthPatchMonintor::StartWatch
    QarthPatchMonintor::WatchPackage: /data/hotpatch/fwkhotpatch/
    QarthPatchMonintor::CheckAndWatchPatch: /data/hotpatch/fwkhotpatch/io.github.nmpribeiro.androidapp
    QarthPatchMonintor::CheckAndWatchPatch: /data/hotpatch/fwkhotpatch/all
    QarthPatchMonintor::Run
I/HwApiCacheMangerEx: apicache path=/storage/emulated/0 state=mounted key=io.github.nmpribeiro.androidapp#10178#256
I/eiro.androidap: QarthPatchMonintor::Reading
    QarthPatchMonintor::CheckNotifyEvent
    QarthPatchMonintor::CheckNotifyEvent before read
I/HwApiCacheMangerEx: apicache path=/storage/emulated/0 state=mounted key=io.github.nmpribeiro.androidapp#10178#0
I/AwareBitmapCacher: init processName:io.github.nmpribeiro.androidapp pid=30377 uid=10178
V/ActivityThread: callActivityOnCreate
E/AwareLog: AtomicFileUtils: readFileLines file not exist: android.util.AtomicFile@fdce1b7
V/HwWidgetFactory: : successes to get AllImpl object and return....
I/OverScrollerOptimization: start init SmartSlideOverScroller and get the overscroller config
    get the overscroller config
W/WindowDecorActionBar: should not do the transition or the transition anim is null or it is running or the mContainer view is null or mContainer view hasn't been drawn to screen 
I/hello-world: rust_jsx_app: Will now wait for native window
    rust_jsx_app: Waiting for NativeScreen
D/ActivityThread: add activity client record, r= ActivityRecord{36f0754 token=android.os.BinderProxy@66c03de {io.github.nmpribeiro.androidapp/android.app.NativeActivity}} token= android.os.BinderProxy@66c03de
D/HiTouch_PressGestureDetector: onAttached, package=io.github.nmpribeiro.androidapp, windowType=1, mHiTouchRestricted=false
I/hello-world: rust_jsx_app: NativeScreen Found:Some(NativeWindow { ptr: 0x7e3e176010 })
    rust_jsx_app: Proceeding after native window
    rust_jsx_app: Inside RUN!
    rust_jsx_app: Inner Size: 0 x 0
    rust_jsx_app: Outer Size: 0 x 0
D/vulkan: searching for layers in '/data/app/io.github.nmpribeiro.androidapp-_bM7KTd-xtIzknMgMxjFjg==/lib/arm64'
    searching for layers in '/data/app/io.github.nmpribeiro.androidapp-_bM7KTd-xtIzknMgMxjFjg==/base.apk!/lib/arm64-v8a'
I/hello-world: wgpu_hal::vulkan::instance: Unable to find extension: VK_EXT_debug_utils
W/hello-world: wgpu_hal::vulkan::instance: Unable to find layer: VK_LAYER_KHRONOS_validation
W/eiro.androidapp: type=1400 audit(0.0:27976): avc: denied { map } for pid=30377 path="/dev/__properties__/u:object_r:huawei_perf_persist_public_read_prop:s0" dev="tmpfs" ino=2822 scontext=u:r:untrusted_app:s0:c178,c256,c512,c768 tcontext=u:object_r:huawei_perf_persist_public_read_prop:s0 tclass=file permissive=0
D/hello-world: wgpu_hal::gles::egl: Client extensions: [
        "EGL_EXT_client_extensions",
        "EGL_KHR_platform_android",
        "EGL_ANGLE_platform_angle",
        "EGL_ANDROID_GLES_layers",
    ]
I/hello-world: wgpu_hal::gles::egl: Using default platform
I/iGraphics: [0020080c] pn: io.github.nmpribeiro.androidapp, p: 30377
    [0030080c] no spt app: io.github.nmpribeiro.androidapp
I/HwViewRootImpl: removeInvalidNode jank list is null
I/hello-world: wgpu_hal::gles::egl: Display vendor "Android", version (1, 4)
D/hello-world: wgpu_hal::gles::egl: Display extensions: [
        "EGL_KHR_get_all_proc_addresses",
        "EGL_ANDROID_presentation_time",
        "EGL_KHR_swap_buffers_with_damage",
        "EGL_ANDROID_get_native_client_buffer",
        "EGL_ANDROID_front_buffer_auto_refresh",
        "EGL_ANDROID_get_frame_timestamps",
        "EGL_EXT_surface_SMPTE2086_metadata",
        "EGL_EXT_surface_CTA861_3_metadata",
        "EGL_KHR_image",
        "EGL_KHR_image_base",
        "EGL_EXT_image_gl_colorspace",
        "EGL_KHR_gl_colorspace",
        "EGL_KHR_gl_texture_2D_image",
        "EGL_KHR_gl_texture_cubemap_image",
        "EGL_KHR_gl_renderbuffer_image",
        "EGL_KHR_fence_sync",
        "EGL_KHR_create_context",
        "EGL_KHR_config_attribs",
        "EGL_KHR_surfaceless_context",
        "EGL_EXT_create_context_robustness",
        "EGL_ANDROID_image_native_buffer",
        "EGL_KHR_wait_sync",
        "EGL_ANDROID_recordable",
        "EGL_KHR_partial_update",
        "EGL_EXT_pixel_format_float",
        "EGL_KHR_mutable_render_buffer",
        "EGL_EXT_protected_content",
        "EGL_IMG_context_priority",
        "EGL_KHR_no_config_context",
    ]
I/hello-world: wgpu_hal::gles::egl: 	Trying presentation
    wgpu_hal::gles::egl: 	EGL context: +debug KHR
    wgpu_hal::gles::egl: 	EGL context: +robust access EXT
I/hello-world: wgpu_hal::gles::egl: 	EGL context: +surfaceless
    wgpu_hal::gles::egl: 	EGL surface: +srgb khr
I/hello-world: rust_jsx_app: Surface
I/RustStdoutStderr: thread '<unnamed>' panicked at 'Cannot get the native window, it's null and will always be null before Event::Resumed and after Event::Suspended. Make sure you only call this function between those events.', /Users/nmpribeiro/.cargo/git/checkouts/winit-1e8677f4a4789b6b/70b2bcd/src/platform_impl/android/mod.rs:594:13
    stack backtrace:
I/RustStdoutStderr: note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
I/hello-world: wgpu_core::hub: Dropping Global
D/AwareBitmapCacher: handleInit switch not opened pid=30377
```

**Comment by Gordon-F** [comment]
@nmpribeiro 

```
I/RustStdoutStderr: thread '<unnamed>' panicked at 'Cannot get the native window, it's null and will always be null before Event::Resumed and after Event::Suspended. Make sure you only call this function between those events.', /Users/nmpribeiro/.cargo/git/checkouts/winit-1e8677f4a4789b6b/70b2bcd/src/platform_impl/android/mod.rs:594:13
```

Workaround:
https://github.com/Gordon-F/gfx/blob/bf7bdc12f93485616f6dd06dd574c2e7afe99e16/examples/quad_android/src/lib.rs#L8-L19

**Comment by nmpribeiro** [comment]
@Gordon-F , I have that if you look to the logs...
```
I/hello-world: rust_jsx_app: Will now wait for native window
    rust_jsx_app: Waiting for NativeScreen
D/ActivityThread: add activity client record, r= ActivityRecord{36f0754 token=android.os.BinderProxy@66c03de {io.github.nmpribeiro.androidapp/android.app.NativeActivity}} token= android.os.BinderProxy@66c03de
D/HiTouch_PressGestureDetector: onAttached, package=io.github.nmpribeiro.androidapp, windowType=1, mHiTouchRestricted=false
I/hello-world: rust_jsx_app: NativeScreen Found:Some(NativeWindow { ptr: 0x7e3e176010 })
    rust_jsx_app: Proceeding after native window 
```

It seems to be either not binding to the correct window or loosing it without me touching it. Regardless, seems a bug.

**Comment by nmpribeiro** [comment]
```
I/hello-world: wgpu_hal::gles::egl: 	EGL context: +surfaceless
    wgpu_hal::gles::egl: 	EGL surface: +srgb khr
    rust_jsx_app: Will now wait for native window
    rust_jsx_app: App started. Waiting for NativeScreen
    rust_jsx_app: NativeScreen Found:Some(NativeWindow { ptr: 0x7dea55a010 })
    rust_jsx_app: Proceeding after native window
    rust_jsx_app: Surface
I/RustStdoutStderr: thread '<unnamed>' panicked at 'Cannot get the native window, it's null and will always be null before Event::Resumed and after Event::Suspended. Make sure you only call this function between those events.', /Users/nmpribeiro/.cargo/git/checkouts/winit-1e8677f4a4789b6b/70b2bcd/src/platform_impl/android/mod.rs:594:13
    stack backtrace:
I/RustStdoutStderr:    0:       0x7de59a709c - <unknown>
       1:       0x7de59c2168 - <unknown>
I/RustStdoutStderr:    2:       0x7de59a5464 - <unknown>
       3:       0x7de59a8c54 - <unknown>
       4:       0x7de59a87a0 - <unknown>
       5:       0x7de59a9288 - <unknown>
I/RustStdoutStderr:    6:       0x7de5482608 - <unknown>
       7:       0x7de5482140 - <unknown>
       8:       0x7de5482560 - <unknown>
I/RustStdoutStderr:    9:       0x7de547c59c - <unknown>
      10:       0x7de547e008 - <unknown>
      11:       0x7de5417740 - <unknown>
      12:       0x7de54169c0 - <unknown>
I/RustStdoutStderr:   13:       0x7de5413560 - <unknown>
      14:       0x7de54185ec - <unknown>
      15:       0x7de5418780 - <unknown>
I/RustStdoutStderr:   16:       0x7de5419b7c - <unknown>
      17:       0x7de5410a70 - <unknown>
      18:       0x7de5419398 - <unknown>
I/RustStdoutStderr:   19:       0x7de5417344 - <unknown>
      20:       0x7de54140cc - <unknown>
      21:       0x7de54113c8 - <unknown>
      22:       0x7de540bb40 - <unknown>
I/RustStdoutStderr:   23:       0x7de540e228 - <unknown>
      24:       0x7de540fef4 - <unknown>
      25:       0x7de54115c8 - <unknown>
      26:       0x7de5419314 - <unknown>
I/RustStdoutStderr:   27:       0x7de5417168 - <unknown>
      28:       0x7de54153dc - <unknown>
      29:       0x7de54154c4 - <unknown>
I/RustStdoutStderr:   30:       0x7de5423b74 - <unknown>
      31:       0x7de542b9d8 - <unknown>
      32:       0x7de5427a6c - <unknown>
I/RustStdoutStderr:   33:       0x7de542b988 - <unknown>
      34:       0x7de542b224 - <unknown>
I/RustStdoutStderr:   35:       0x7de542b944 - <unknown>
I/RustStdoutStderr:   36:       0x7de542b18c - <unknown>
      37:       0x7de542bca4 - <unknown>
I/RustStdoutStderr:   38:       0x7de5427914 - <unknown>
      39:       0x7de542a18c - <unknown>
I/RustStdoutStderr:   40:       0x7de59aad64 - <unknown>
I/RustStdoutStderr:   41:       0x7ed92df6f4 - <unknown>
I/RustStdoutStderr:   42:       0x7ed92820ec - <unknown>
      43:                0x0 - <unknown>
I/hello-world: wgpu_core::hub: Dropping Global
```

**Comment by nmpribeiro** [comment]
Found a potential root cause for this issue. I have tried to fix the surface not found on startup by adapting my logic to use `Resume`/`Suspended` events. However, only with `winit` `v0.24` that works. If I move to `v0.25` or latest master the behaviour is changed, not triggering that event on startup.

This is what happens with `winit` from my forked repo and ios fix, forked from tag v0.23:
```
winit = { git="https://github.com/nmpribeiro/winit.git", branch="working_android_winit"}
```

While in my android simulator it doesn't work, it is expected that support for OpenGL for android is still lagging behind. Therefore, will move this to `winit` and close this issue.
